diff --git a/commands.c b/commands.c
index 97cadf2..f198efb 100644
--- a/commands.c
+++ b/commands.c
@@ -253,6 +253,16 @@ bool cg_change_contrast(arg_t d)
 	return change_color_modifier(d, &img.contrast);
 }
 
+bool cg_toggle_winbg(arg_t _)
+{
+	win.win_bg.pixel ^= ~0;
+	win.win_bg.red   ^= ~0;
+	win.win_bg.green ^= ~0;
+	win.win_bg.blue  ^= ~0;
+	img.dirty = true;
+	return true;
+}
+
 bool ci_navigate(arg_t n)
 {
 	if (prefix > 0)
diff --git a/commands.h b/commands.h
index 4e694f0..8552215 100644
--- a/commands.h
+++ b/commands.h
@@ -21,6 +21,7 @@ bool cg_switch_mode(arg_t);
 bool cg_toggle_bar(arg_t);
 bool cg_toggle_fullscreen(arg_t);
 bool cg_toggle_image_mark(arg_t);
+bool cg_toggle_winbg(arg_t);
 bool cg_unmark_all(arg_t);
 bool cg_zoom(arg_t);
 /* image mode */
@@ -67,6 +68,7 @@ bool ct_select(arg_t);
 #define g_toggle_bar { cg_toggle_bar, MODE_ALL }
 #define g_toggle_fullscreen { cg_toggle_fullscreen, MODE_ALL }
 #define g_toggle_image_mark { cg_toggle_image_mark, MODE_ALL }
+#define g_toggle_winbg { cg_toggle_winbg, MODE_ALL }
 #define g_unmark_all { cg_unmark_all, MODE_ALL }
 #define g_zoom { cg_zoom, MODE_ALL }
 
diff --git a/config.def.h b/config.def.h
index 5d4e97a..4b1cf20 100644
--- a/config.def.h
+++ b/config.def.h
@@ -96,6 +96,7 @@ static const keymap_t keys[] = {
 	{ 0,            XK_Return,        g_switch_mode,        None },
 	{ 0,            XK_f,             g_toggle_fullscreen,  None },
 	{ 0,            XK_b,             g_toggle_bar,         None },
+	{ 0,            XK_t,             g_toggle_winbg,       None },
 	{ ControlMask,  XK_x,             g_prefix_external,    None },
 	{ 0,            XK_g,             g_first,              None },
 	{ 0,            XK_G,             g_n_or_last,          None },
